backend/proxy/secure_socks_proxy.go

```diff
@@ -67,6 +68,19 @@ type ClientCfg struct {
 	AllowInsecure bool
 }
 
+// Hash is used as part of the plugin instance cache key.
+// The ClientKeyVal and ClientCertVal differ between different Hosted Grafana instances,
+// so we include them in the hash key to avoid false positives in `NeedsUpdate`.
+func (c *ClientCfg) Hash() string {
+	if c == nil {
+		return ""
+	}
+	h := fnv.New64a()
```

Comment:
Have you thought about how adding this affects performance? I had originally thought that maybe specifying an ID for the proxy config would be better (HG writes the same for instances of the same stack), and then we just use that as part of the instance key instead?

Reply:
> Have you thought about how adding this affects performance?
I have, yeah. Fnv is specifically designed for speed, and this will only be triggered on network requests, the latency of which should dominate. Just to be sure, though, now I've benchmarked it as well. It's a bit hard to get accurate results because the setup takes longer than the function, but it takes less than 500 ns per call to `ProxyHashFromContext` on my laptop, and only about half of that is due to fnv itself. That seems acceptable to me, for a network-driven code path, but please let me know if you disagree.

>  I had originally thought that maybe specifying an ID for the proxy config would be better
That does seem cleaner, but I don't think it's something I can tackle at the moment, and my understanding from @dafydd-t is that it's unlikely to get prioritized on that side very soon. This issue has already caused significant problems for at least one customer, and I think may have contributed to another recent escalation, so I'd like to fix it sooner than later.

Reply:
I'm happy to drive that work if we establish that it's a better path to go down - but I'll wait for @dafydd-t to weigh in. 

Also - is it fine to just hash the cert key or is the value also required?

Reply:
> Also - is it fine to just hash the cert key or is the value also required?

I think you could take just the client key or the client cert - they are generated together. In fact, the cert contains a signature from the key, so just hashing the cert should be sufficient.

I think I would add the Root CAs to the hash, though, to handle the case where we have to rotate CA keys, we want the datasource API to refresh too.

Reply:
It is unlikely, but we might avoid some pain in the future if we also include ProxyAddress and ServerName.

Reply:
> I had originally thought that maybe specifying an ID for the proxy config would be better (HG writes the same for instances of the same stack)

To me, this would end up  with HGAPI generating generating a hash and then passing it down to the proxy config, instead of calculating it here. We couldn't use a static number as we'd need to update it whenever we generate a new client key pair (which we do a lot), and if we wanted a ever-increasing id / UUID, we'd have to keep that state somewhere. We have no need for a number like that in HGAPI.

A middle ground I guess would be to store the hash of the clientCfg in the config itself? Then we'd only have to calculate the hash for a given clientCfg once.

I'm happy with the current approach, or caching the hash in ClientCfg, but I don't think its worth the work to move to a unique ID in HGAPI personally. Of course I'm open to discussing it, I might have missed something.

Reply:
Thanks for weighing in @dafydd-t!

>but I don't think its worth the work to move to a unique ID in HGAPI personally

That's fair 👍 What I suggested ended up sounding more complicated in reality, but your suggestion of

>store the hash of the clientCfg in the config itself

sounds good to me. I had imagined originally HG would do this, but Grafana can just do this once and then pass it with each request [here](https://github.com/grafana/grafana/blob/main/pkg/services/pluginsintegration/pluginconfig/request.go#L75).

I'm not dead set on any approach, but I do like the idea of only calculating this once and keeping the cache key check very simple.

Reply:
I like this idea of hashing once! Simple and efficient. I can open a new PR for that.

Reply:
> I think I would add the Root CAs to the hash, though, to handle the case where we have to rotate CA keys, we want the datasource API to refresh too.

We don't actually need to worry about this, I realized as I was putting this together: the full config is cached with the datasource instance, and we discard the instance if anything has changed. In other words, adding just the cert hash to the key will eliminate the false positive from `NeedsUpdate` but doesn't introduce a risk of a false negative.

Reply:
Nice, thanks for explaining.

Reply:
@njvrzm are you going to remove the ClientKeyVal from the hash? I'll approve anyway, but I don't think its needed.

Reply:
Yes, I'll do that. I'll need to rework this a little bit for use from grafana proper, too.

Reply:
The SDK changes needed should be really small in fact as you'll just need to add a new field for the hash [here](https://github.com/grafana/grafana-plugin-sdk-go/blob/main/backend/proxy/secure_socks_proxy.go#L55) and do the heavy lifting in `grafana/grafana`!

Reply:
For my understanding, why is it better to have grafana/grafana calculate the hash? I was imagining that the ClientCfg would have a public Hash() method and a private hash property, and we could calculate the hash and store it in the private field if it wasn't set already.  I feel like this would keep the logic in one place, but I think I am missing something?

Reply:
The hash would need to be calculated on each plugin request, by the plugin - whereas having Grafana do it means we do it once when it's read and subsequently sent with every request.

Reply:
> Is there a reason the "calculated once" value cannot be stored in the ClientCfg
We can't store the hash on the ClientCfg because that's created anew for every request, but the GrafanaCfg where I'm storing the hash now persists for the lifetime of the datasource instance.

Reply:
Thanks for explaining!

Reply:
Haha sorry to be a pain but this is still different than what I had originally suggested (plugin still has to do this itself at request time) 🙈  As long as it works as expected and you're both happy, then it's alright though. I'd just like to take it for a spin before merging!

Reply:
I understand that this is different, but I don't understand why it's importantly different that the hash happens one time with the first request an instance receives as opposed to one time when the instance is created.

Reply:
The idea mentioned before was that Grafana would do it only once and then would be sent for every request - otherwise with other solutions it has to be recalculated (when it never actually changes). I just thought that was ideal, but can see the concern about spreading code between grafana/grafana and the SDK. There could be some effort spent to make the code could live in the SDK though potentially (a function to create the map with pdc config and automatically populate the hashing field for example).

>but the GrafanaCfg where I'm storing the hash now persists for the lifetime of the datasource instance.

This isn't true unfortunately (is tied to request and not instance) and therefore the hash is always recalculated.

Reply:
> This isn't true unfortunately (is tied to request and not instance) and therefore the hash is always recalculated.

🤦🏻 I think I misunderstood something you said earlier, and didn't follow the code trail back far enough. My apologies. Sigh.

But I'd still really prefer to keep this in the sdk, if possible. Two options I see for that:
1. Accept the 1-2 microsecond latency hit on requests. 
2. Cache the cert -> hash mapping on the instance provider instead of in the `GrafanaCfg`.

Now that I think about it, though, if we're going to do 2, we might as well skip the hash: just use the client key (since it's smaller) as part of the instance cache key. In fact we could just use the last, say, four bytes of it - the beginning is the same for all of them. The key is encoded in base64, so that gets us an acceptably low collision rate.

So, my final proposal:
3. Add the last 4 bytes of the client key value to the instance cache key.

Going to pause for one last round of feedback 😅 but I hope this is acceptable to everyone.

Reply:
Well - I went ahead with that approach, since it's getting late on Friday for me. Let me know what you think.

Reply:
> Add the last 4 bytes of the PDC client key value to the instance cache key.

The base64-decoded key is in PEM format (at least it is on file), so the last line is always the same. I'd recommend something a bit longer

Reply:
For the record, @dafydd-t, you're totally right, that's also true for the key contents passed in, so I've opened #1157 to fix that.